### PR TITLE
LibWeb+RequestServer: Add and implement XMLHttpRequest::timeout

### PR DESCRIPTION
**Pull request by kennethmyhra**

### CODE REVIEW

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp [hunk 1]
```diff
@@ -693,4 +694,22 @@ DOM::ExceptionOr<void> XMLHttpRequest::override_mime_type(String const& mime)
 
     return {};
 }
+
+// https://xhr.spec.whatwg.org/#the-timeout-attribute
```
Comment (linusg) [inline Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp:698]:
```suggestion
// https://xhr.spec.whatwg.org/#ref-for-dom-xmlhttprequest-timeout%E2%91%A2
```

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp [hunk 2]
```diff
@@ -693,4 +694,22 @@ DOM::ExceptionOr<void> XMLHttpRequest::override_mime_type(String const& mime)
...
+    if (global_object.class_name() == "WindowObject" && m_synchronous)
+        return DOM::InvalidAccessError::create("Use of XMLHttpRequest's timeout attribute is not supported in the synchronous mode in window context.");
+
+    // 2. Set this’s timeout to the given value.
+    m_timeout = timeout;
+
+    return {};
+}
+
+// https://xhr.spec.whatwg.org/#the-timeout-attribute
```
Comment (linusg) [inline Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp:713]:
```suggestion
// https://xhr.spec.whatwg.org/#dom-xmlhttprequest-timeout
```

File: Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp [hunk 3]
```diff
@@ -693,4 +693,22 @@ DOM::ExceptionOr<void> XMLHttpRequest::override_mime_type(String const& mime)
...
+size_t XMLHttpRequest::timeout() const { return m_timeout; }
```
Comment (linusg) [inline Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp:714]: Missing newline before namespace

File: Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp [hunk 4]
```diff
@@ -250,6 +250,12 @@ void ResourceLoader::load(LoadRequest& request, Function<void(ReadonlyBytes, Has
             return;
         }
         m_active_requests.set(*protocol_request);
+
+        protocol_request->on_progress = [request, &protocol_request, timeout](Optional<u32>, u32) {
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp:254]: What if there never is progress in the first place though? I feel like this is something that should be handled by a timer instead...

Comment (kennethmyhra) [reply Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp]: Agree. I've added a timer into the mix. Felt a bit hackish the way I went around to make it work, let me know if you think there's some other more proper way I could go about this.

File: Userland/Libraries/LibWeb/Loader/ResourceLoader.h [hunk 5]
```diff
@@ -98,6 +99,7 @@ class ResourceLoader : public Core::Object {
     HashTable<NonnullRefPtr<ResourceLoaderConnectorRequest>> m_active_requests;
     NonnullRefPtr<ResourceLoaderConnector> m_connector;
     String m_user_agent;
+    NonnullRefPtr<Core::Timer> m_timer = Core::Timer::create_single_shot(0, [] {});
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Loader/ResourceLoader.h:102]: hmm, perhaps we could make this a RefPtr and only create a timer if the request's timeout is actually set?

File: Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp [hunk 6]
```diff
@@ -287,11 +299,11 @@ void ResourceLoader::load(LoadRequest& request, Function<void(ReadonlyBytes, Has
         error_callback(not_implemented_error, {});
 }
 
-void ResourceLoader::load(const AK::URL& url, Function<void(ReadonlyBytes, HashMap<String, String, CaseInsensitiveStringTraits> const& response_headers, Optional<u32> status_code)> success_callback, Function<void(String const&, Optional<u32> status_code)> error_callback)
+void ResourceLoader::load(const AK::URL& url, Function<void(ReadonlyBytes, HashMap<String, String, CaseInsensitiveStringTraits> const& response_headers, Optional<u32> status_code)> success_callback, Function<void(String const&, Optional<u32> status_code)> error_callback, Optional<i64> timeout)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp:302]: I don't really like that this uses an i64 and the other place uses a size_t, which is 32 bits on x86. Can we stick to one?

File: Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp [hunk 7]
```diff
@@ -137,10 +137,21 @@ static void emit_signpost(String const& message, int id)
...
+    RefPtr<ResourceLoaderConnectorRequest> protocol_request;
+    if (timeout.has_value() && timeout.value() > 0) {
+        m_timer->on_timeout = [request, &protocol_request] {
+            dbgln("ResourceLoader: Stop the request!");
+            protocol_request->stop();
+        };
+        m_timer->set_interval(timeout.value());
+        m_timer->restart();
+    } else {
+        m_timer->stop();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp:153]: If we only allocate the timer above, we can probably drop this branch

**Review by linusg — COMMENTED** [review]
Looks good! A few more thoughts :)

**Comment by kennethmyhra** [comment]
Thanks! Made `m_timer` a `RefPtr`. Moved both `size_t` and `i64` to `i32`

**Review by linusg — APPROVED** [review]
Thanks, LGTM!

**Review by linusg — CHANGES_REQUESTED** [review]
Hold on - IDL `unsigned long` is `u32`, not `i32`:

https://github.com/SerenityOS/serenity/blob/2fa907b31a4172b7cbe3cc67e63dc590e9f3a23b/Meta/Lagom/Tools/CodeGenerators/LibWeb/WrapperGenerator/IDLGenerators.cpp#L91-L92

**Comment by kennethmyhra** [comment]
Ah yes of course, the `unsigned long` was why I had 'XMLHttpRequest::timeout()' and `XMLHttpRequest::set_timeout()` as `size_t` initially. Moved all of it to u32 now.

Thanks for the swift reviews, Linus!

**Pull request merged**